dapps/src/router.rs

```diff
@@ -109,6 +112,7 @@ impl http::RequestMiddleware for Router {
 				=>
 			{
 				trace!(target: "dapps", "Resolving to 404.");
+				refresh.unwrap().refresh_local_dapps();
```

Comment:
`self.endpoints` can be empty so you should unwrap.
It should be just `self.endpoints.as_ref().map(|endpoints| endpoints.refresh_local_dapps())`.

Even though the dapps are refreshed we should still re-try resolving the request.
Most likely a recursive call should be done here with some additional parameter preventing a refresh again.